### ISSUE TITLE
Hight Akku drain since last update.

### ISSUE THREAD

**Issue by Mannshoch** [andstatus/andstatus#340]
Last update from f-droid got me a hot Mobil phone and consume more then 50% of the energy.
I had to uninstall it.

**Comment by yvolk** [comment]
Maybe some process got stuck during upgrade and you only had to restart your device?!
Please try to reinstall AndStatus again.
In a case of the same high energy consumption please share with me the first entry of the "Commands in a queue" list. It contains information on background tasks.
It should contain something like this:
```
Tasks:
Total launched: 25; running: 1. Skipped: 0
R 1. TimelineActivity133 on LONG_UI; age 4sec; RUNNING 4sec; instanceId=137; org.andstatus.app.LoadableListActivity$AsyncLoader@b145d5c1

Thread pools:
SYNC: java.util.concurrent.ThreadPoolExecutor@e57173d[Running, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
FILE_DOWNLOAD: java.util.concurrent.ThreadPoolExecutor@a57ab32[Running, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
QUICK_UI: java.util.concurrent.ThreadPoolExecutor@8716200[Running, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
LONG_UI: java.util.concurrent.ThreadPoolExecutor@22ea39[Running, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 0]
DEFAULT: java.util.concurrent.ThreadPoolExecutor@351487e[Running, pool size = 2, active threads = 0, queued tasks = 0, completed tasks = 2]

ImageCaches:
Avatars size:165888 of 8201749 bytes, hits:34, misses:18, hitRate:65%
Attached images size:0 of 41008744 bytes, hits:0, misses:0
Styled drawables: 0
Memory: available 588247040 of 1922715648
```

**Comment by Mannshoch** [comment]
Maybe I got the cause of the Problem. 
I Installed Xprivacy and maybe Andstatus got a Problem because I was not Open it and setup the allowed rights, so the Background Process got a problem.

**Comment by Mannshoch** [comment]
It seems I was wrong. Also with the settings from xPrivacy corrected I had a high Akku drain.

```
T; empty
Erstellt vor 3 Sekunden

Tasks:
Total launched: 84; running: 1. Skipped: 0
R 1. HeartBeat 0; org.andstatus.app.service.MyService$HeartBeat on SYNC; age 0sec; RUNNING 0sec, cancelled; instanceId=536; org.andstatus.app.service.MyService$HeartBeat@6d19fe6c

Thread pools:
SYNC: java.util.concurrent.ThreadPoolExecutor@395ec946[Running, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
FILE_DOWNLOAD: java.util.concurrent.ThreadPoolExecutor@eb1b437[Running, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
QUICK_UI: java.util.concurrent.ThreadPoolExecutor@1cada4[Running, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]
LONG_UI: java.util.concurrent.ThreadPoolExecutor@39f6170d[Running, pool size = 1, active threads = 0, queued tasks = 0, completed tasks = 1]
DEFAULT: java.util.concurrent.ThreadPoolExecutor@9dbf6c2[Running, pool size = 2, active threads = 0, queued tasks = 0, completed tasks = 2]

ImageCaches:
Avatars size:156672 of 5346591 bytes, hits:83, misses:17, hitRate:83%
Attached images size:0 of 26732954 bytes, hits:0, misses:0
Memory: available 264269824 of 828674048
```

**Comment by yvolk** [comment]
From your report I don't see anything that could cause an energy consumption...

**Comment by Yowlen** [comment]
I think it has to do with the screen turning on every so often to sync. My device has been doing it for a while now, too, but I was unable to test to make sure it wasn't some other app until now. I'm not sure what the requirements are when using the built-in Android account sync, but I don't think turning the screen on is necessary for most devices. I'm aware some devices need it in order to speed up downloads (my device is one of them), but for a periodic sync like this, the slower screen off speed will work just fine.

To accommodate those that do want the screen on (busy timelines might need this), I'd suggest adding an option in the settings to toggle the screen wakelock (default off) with a description saying "If you have trouble syncing in the background, try enabling this."

Once implemented, then the original poster can give it another shot to see if their problem is solved.

**Comment by yvolk** [comment]
@Yowlen I never heard about turning screen on during Android syncing, and I never noticed this effect on my devices. At least AndStatus doesn't do anything explicitly except it prevents device from sleeping during a sync, which is normal.
So I think that screen behaviour is (and should be) controlled by some System-wide settings, not inside AndStatus.

**Comment by Yowlen** [comment]
Sorry for the red herring. I'm trying to debug my screen wakelock issue without root, so it makes it harder thanks to Google making Android's BatteryStats API a privileged|system one. That's one thing from Jellybean I miss...

Anyway, after using Force Stop on AndStatus, I can see it wasn't this app that caused it. In fact, as I look at it more, I noticed it didn't even sync up with AndStatus' sync schedule. It seemed to occur once every hour or two, while I have AndStatus set to every 15 mins. Your system-wide setting thing sparked an idea for me, though, so thanks for that. I forgot I had messed with the wifi settings to try to help with another app I was having issues with. With any luck, my issue might be solved now. :)

@Mannshoch I hope you discover the cause of your issue. And again, sorry for the false lead. :/

**Comment by Yowlen** [comment]
Well, I bought a new phone a few days ago (LG Leon HR345 running 5.1.1), and I noticed it drains the battery rather quickly, so I did some digging, and it turns out the original poster does have some merit to their claim. I measured the battery drain with AndStatus running overnight, and again the next night with AndStatus not running (frozen/disabled). The results:
With AndStatus: 19% drain over 8 hours
Without AndStatus: 12% drain over 8 hours

I also noticed something else. AndStatus itself isn't the cause of the drain, but rather, some system processes. I still have some avenues left to explore to narrow down the cause, but I suspect it has to do with the syncing, since that was the major revamp being done in the app at the time.

My idea on the cause (for information, & in case someone else can test quicker than me):
AndStatus seems to not fully stop (not sure why), and notifications sometimes show more often than the sync interval, which leads me to suspect a secondary sync is in place. (Perhaps the legacy system is still in the code, and is being called when it shouldn't be?)

I will disable certain sync systems to try to figure out specific areas that cause this drain. My current sync settings are:
- Frequency: every 15 mins (I sometimes get notifications within 5 mins of each other, though that hasn't occurred yet for me since the 23.00 release.)
- Sync while using AndStatus: On
- Sync indicator in a Timeline: On
- Sync after message was sent: Off
- Connection timeout: 30s

I doubt that any but the first option would make a difference, and I'm not sure I can do much more experimentation, but I'll keep trying in case I discover something useful. :)

**Comment by yvolk** [comment]
@Yowlen I agree that syncing could be the most energy consuming process.
Some statements for you to check:
1. If you want to have full information about what AndStatus do around the clock, you should turn saving logs to files and inspect that logs. See https://github.com/andstatus/andstatus/issues/225 for info about logging.
2. Inside AndSta

**Comment by Yowlen** [comment]
1. Thanks for the reminder about log files. Can't believe I forgot that one. And just a couple days ago, I had created a script to help my parents with their digital camera - with logging in case something went wrong. lol
I'll get back to you once I have some logs to analyze.
2. Something else I thought of:
Does AndStatus wake itself for syncs, or does it dutifully wait for the system to tell it to wake up? Just making sure it's the latter, since the former would cause issues on devices with wifi saving on the default settings. It will often disconnect from wifi when the screen is off, which sometimes causes sync fails and/or other app irregularities like my FolderSync thing from before. I would assume that the system would detect the lack of connection and not wake AndStatus during a sync in this case, but if AndStatus wakes itself, it'd be extremely wasteful.
3. Another thought:
During the sync, how many items can fail? Is there a "critical failure" point that stops the sync if X number of items fail? If so, does the system get notified? I've seen sometimes that some of my Google syncs fail if they suddenly get interrupted via an icon in the system's Settings > Accounts area. I have a game that's very demanding on RAM, so it sometimes cuts off other apps in the middle of syncing.

**Comment by yvolk** [comment]
2. AndStatus doesn't wake itself, as you expected. So every new bunch of log files at night should start with some sync adapter...

3. During each sync several timelines of one account are synced. As I remember, if one of timelines fails, AndStatus reports an error to Android sync, Ind that Android sync, not AndStatus, decides, if and when to retry.
BUT during sync AndStatus may start executing other queued tasks also, e.g. avatars and attachments downloading...

**Comment by Yowlen** [comment]
Ah. So the pictures/avatars aren't part of the system sync. From the sound of it, the system sync first syncs timeline text, then calls a script to start the Command Queue for the rest. Wouldn't it be more efficient to have the entire Command Queue  in the sync, then when the user calls for a manual refresh, send the call to the system to manually update the AndStatus account?

Also, does the Command Queue re-sync the timeline? If so, that could be a potential cause for the drain, since it would essentially result in a loop without a specific check to make sure no loops occur.

Either way, the above idea of mine would probably work better. It could allow for more possibilities, too. Individual system sync items for timelines, and perhaps separate items for pictures & avatars so users could sync avatars and just open pictures on-demand in the browser to save bandwidth/battery during syncs.

**Comment by Yowlen** [comment]
Another idea formed in my head today:
I've noticed AndStatus shows the number of refreshes on Twitter's API. It seems to show the number done, and not the number left. I know Twitter's API has a 15 refreshes in 15 minute limit per account. And I realized today that since the 22.00-ish release that changed the sync from an app-based one to a system account one, that number has increased drastically. If I recall, the number pre-22.00 was around 2-3/15. But now, when I first check AndStatus in the morning after waking up, the number is consitently 11+/15. And even more disturbing is the recent thing Twitter's been doing since I got my 2nd device - blocking access to the mobile site because I've apparently "reached my daily limit" for it. After waiting a few minutes to allow the API limit to free up an attempt, I can usually access the site again.

So, since that sync redo release is when the battery drain increased, and this also started happening with it, I can't help but think the two are connected.

**Comment by yvolk** [comment]
@Yowlen It's interesting about your mentioning this: "And I realized today that since the 22.00-ish release that changed the sync from an app-based one to a system account one, that number has increased drastically. If I recall, the number pre-22.00 was around 2-3/15. But now, when I first check AndStatus in the morning after waking up, the number is consitently 11+/15."
Do you mean that Twitter changed their rules some time ago? If yes, why do you mention v.22 of AndStatus, if I understood you correctly?! - I didn't change anything about sync recently...
Please clarify!

**Comment by Yowlen** [comment]
I'm not sure on the version number, but I assumed around version 22.00 you changed something that severely increased battery drain. It was definitely around late February, and I see several commits from that time. I noticed that before that timeframe, AndStatus didn't have a sync page in Android's system settings, and I recall that AndStatus ran in the background all the time to update itself during that time since it didn't rely on the Android sync to let it know when to wake up.

To summarize, the problems mentioned popped up around the same time as the switch to Android's built-in sync system, so I can only assume that's the cause.

**Comment by yvolk** [comment]
AndStatus has its accounts and syncing in Android system settings for years.
Yes, recently I did a lot of changes to AndStatus background activities. In particular, image downloading tasks have now their own thread pool, which allows us to start execution of timeline updates faster. 
So maybe AndStatus loads too long too many attachments and avatars to your device?! By default it does this on WiFi only...

Please try to analyze logs as I suggested above...

**Comment by yvolk** [comment]
Maybe I found a cause of these problems: incorrect ACRA (crash reporting) configuration.
On the page https://github.com/ACRA/acra/wiki/BasicSetup I see the note in Bold:

> NB if you are starting services or other tasks in Application.onCreate, you should test to see if the current process is the :acra process. If it is then you probably don't want to start your service.

The note appeared quite recently and I didn't see it before. Investigating...

**Comment by Yowlen** [comment]
@yvolk Yeah, sorry about the lack of logs. Recently reset my device & forgot to re-enable logging.

I'm glad you found a possible cause. Maybe logs won't be needed anymore. Still, I'll let AndStatus log for a few hours, then check & post the results.

**Comment by Yowlen** [comment]
Well, I have a lot of logs here. I did two sets of them. One is from AndStatus 22.05 (using F-Droid's Archive repo) which I think is the last version before the image cache rework and therefore the last known version without the issue. The other is from AndStatus 24.03, the latest version in the F-Droid repo.

[AndStatus 22.05 Logs.zip](https://github.com/andstatus/andstatus/files/218214/AndStatus.22.05.Logs.zip)
[AndStatus 24.03 Logs.zip](https://github.com/andstatus/andstatus/files/218219/AndStatus.24.03.Logs.zip)

Unrelated note: GitHub's mobile site sucks. Can't filter issues, can't attach files to comments. Gotta use the desktop version for that which is hard to read due to tiny print.
